### ISSUE TITLE
Incorrect `main` function requirement in workspace examples module

### ISSUE THREAD

**Issue by Jonas-Heinrich** [rust-analyzer/rust-analyzer#10858]
# Incorrect `main` function requirement in examples module when using workspaces

Hi,

upon creating a standalone module file (as [changed in edition 2018](https://doc.rust-lang.org/edition-guide/rust-2018/path-changes.html)) within the `examples` directory of a workspace, I get displayed an erroneous error message for the entire file.
This is an issue with rust-analyzer in VS Code, as the code compiles and executes fine.

## Steps to Reproduce

I tried to keep the reproduction setup to a minimum.

### 1. Create new rust project

```bash
cargo new mod-bug
```

### 2. Create the project layout

```
src/
    lorem/
        examples/
            main.rs
            sub.rs
        Cargo.toml
Cargo.toml
[...]
```

### 3. Fill files

```rust
// File main.rs
mod sub;
use crate::sub::hello;

fn main() {
    hello();
}
```

```rust
// File sub.rs
pub fn hello() {
    println!("Hello from sub module");
}
```

```toml
# Outer Cargo.toml (`Cargo.toml`)

[workspace]
members = [
    "src/*",
]
```

```toml
# Inner Cargo.toml (`src/lorem/Cargo.toml`)

[package]
name = "lorem"
version = "0.0.1"
authors = ["Zarathustra"]
edition = "2018"

[[example]]
name = "main"
required-features = []
doc = true
```

### 4. Confirm execution

```sh
> cargo run --package lorem --example main
   Compiling lorem v0.0.1 ([...]/mod-bug/src/lorem)
    Finished dev [unoptimized + debuginfo] target(s) in 0.55s
     Running `target/debug/examples/main`
Hello from sub module
```

(works as intended)

## Issue

When viewing the `sub.rs` file in VS Code with rust-analyzer enabled, the entire file is marked with red squiggly lines indicating an error, however it is a valid setup and does execute.

![image](https://user-images.githubusercontent.com/14094195/143444997-83ab7c89-8c21-4546-a3f5-cbf4650d58fe.png)

This is obviously annoying for viewing or editing.
The issue also exists when changing the rust edition in the outer `Cargo.toml` file to 2021.

## Software Versions

- OS: openSUSE Leap 15.3
- rustc 1.56.1 (59eed8a2a 2021-11-01)
- cargo 1.56.0 (4ed5d137b 2021-10-04)
- Rust Edition 2018, 2021
- VS Code 1.62.3
- rust-analyzer v0.2.826

**Comment by jhgg** [comment]
I don't believe this is an issue, each ".rs" file within the `examples` directory of a crate is considered an example build target, as illustrated by:


```
❯ cargo run --package lorem --example
error: "--example" takes one argument.
Available examples:
    main
    sub

```

rust-analyzer is using the result of `cargo check --examples` in this situation, and is considering `sub` as its own target:

```
❯ cargo check --examples
    Checking lorem v0.0.1 (/home/jake/code/test3/mod-bug/src/lorem)
error[E0601]: `main` function not found in crate `sub`
 --> src/lorem/examples/sub.rs:2:1
  |
2 | / pub fn hello() {
3 | |     println!("Hello from sub module");
4 | | }
  | |_^ consider adding a `main` function to `src/lorem/examples/sub.rs`

For more information about this error, try `rustc --explain E0601`.
error: could not compile `lorem` due to previous error
```

**Comment by lnicola** [comment]
Yeah, here's an example of a crate with multi-file examples: https://github.com/tokio-rs/axum/tree/main/examples.

**Comment by Jonas-Heinrich** [comment]
@jhgg I was unaware of the fact, that each `.rs` file in `examples` is implicitly considered as an example build target. In that case, I follow your argument and agree that its not an issue.

Thank you!
